cilium: encryption, auto-discover interface and subnet

jrfastab opened the pull request:
Resolve numerous issues that have crept into encryption side with EKS, AKS, and GKE environments.

* EKS requires editing config-map with a set of undocumented cilium-internal fields to get working. Instead of exposing these lets learn these from the environment. Patches 1,2, and 3 address this so that subnets and encryption interfaces are learned at init time.
* EKS interface auto-detection was not reliable, fix by patch 3 to add all network facing interfaces. The issue here is EKS uses source based routing and has multiple interfaces where encrypted traffic could arrive. If we don't load our BPF program bpf-network on these interfaces packets may not be marked for decryption and dropped by the stack.
* We attempted to enable Extended sequence support, but missed two bits of information here: (i) the seq# will need to be sync'd on node join and (ii) the encryption overhead is increased. Rather than deal with the somewhat more complicated solution for (i) lets revert for now to make backporting doable. Cilium-cli should make key rotations easy and painless so that ESN is not needed for most cases. Patch 4.
* Patches 5, 6 ensure IPSec configuration is cleaned up. Otherwise we risk having conflicting rules in the datapath when running install, uninstall, install and/or restarting cilium-agent if it comes up with new IP addresses, cidrs. Which in practice is unlikely, but can happen. In CI its commonplace because we reused the nodes across tests.

Tested on EKS using,
* ./cilium install --agent-image quay.io/isovalent/cilium-hubble-fgs --operator-image quay.io/cilium/operator-aws-ci --version latest --encryption
* ./cilium connectivity test
   ✅ 9/9 tests successful (0 warnings)
   
Tested on GKE using
*./cilium install --datapath-mode=tunnel --agent-image quay.io/isovalent/cilium-hubble-fgs --operator-image quay.io/cilium/operator-generic-ci --version latest --encryption
* ./cilium connectivity test

And
* ./cilium install --datapath-mode=tunnel --ipam=cluster-pool --agent-image quay.io/isovalent/cilium-hubble-fgs --operator-image quay.io/cilium/operator-generic-ci --version latest --encryption
* ./cilium connectivity test

All three pass 9/9/ tests successfully. To test transitions between encryption and non-encryption I cycled through above with `install --encryption`, `uninstall`, `install`, `uninstall`, `install --encryption`. This mostly works however with `--ipam=cluster-pool` case in GKE there is some flake where encryption does not come back online correctly in last case and tests fail. I am attempting to track that down now, but don't think it needs to block this series.

jrfastab committed:
cilium: encryptInterface requires multiple interfaces in multi-dev env

When a node is using multiple nodes and traffic is being source routed,
or routed some other way, across multiple netdevs we need to ensure
our bpf_network programs are attached to all interfaces that can receive
pod traffic.

This patch adds support for users to specify multiple interfaces. In
order to do this lift bpf_network (encryption program) compile and load
out of init.sh and do it from go side.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: auto-discovery pod subnets for ENI IPAM

When encryption is enabled with an IPAM using a single IP pool for the
entire cluster instead of per node we need to know those CIDRs. This is
required so we can place a decrypt/encrypt rules that covers the subnet.

In this patch lookup the CIDRs from the routeInfo and use those. Otherwise
users would need to manually modify helm charts, yamls, etc.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: auto-detect network interfaces for IPAMENI case

We need to auto-detect network interfaces in the EKS environment that by
default uses three interfaces eth0, eth1, and eth2. This patch will walk
the list of all devices in the system and attach to any device that is
not a veth or lo device.

This patch resolves the EKS case and leaves other cases for future work.
We could use similar logic with a bit more exclusion logic to handle the
general routing cases as well.

In the process we pull the IPAMENI IPSec configuration into Reinitialize()
path so that if net devices are added/removed we will pick up those at
reinit time. Later work can bring the remaining parts, but that required
more refactoring than I wanted for a bugfix.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: remove rules when encryption is disabled

Rules need to be correctly deleted otherwise its possible if in
the future cilium agent picks up a new IP address and then
re-enables encryption we may have some conflicting entries in
the route table.

Also its just cleaner to have only correct entires in the table
attempting to debug a system with multiple invalid routes is
more difficult than it needs to be.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: Use correct mask on xfrm delete

The mask values used to match stale entries in the XFRM state and policy
no longer matched the values we use. The result is stale entries get
left in the node tables. If subsequently cilium agent picks a new IP
address this can result in conflicts in the xfrm {state|policy} lookup

Fix to use correct mask field.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, insert "skip" rules to avoid ipsec policy

In some configurations of tunnel mode, endpoint routes, etc. An encrypted
packet will be sent through the stack. If the encrypted packet does not
match a fwd policy rule AND it has previously been encrypted without passing
through a kernel boundary that clears the cb[] (on older kernels) or skb
extension (on newer kernels) fields the xfrm policy will miss and the
packet will be dropped.

This worked in the past because we mainly had older kernels where xfrm
stack used cb[] for metadata and we cleared the cb[] fields from BPF
stack. However, we do not clear skb extensions via BPF nor do we have
a BPF helper to do this yet. So in order to deal with this case we
add fwd xfrm policy rules so that when the routing stack detects the
skb extensions for xfrm and attempts to do a policy lookup a policy
is found and the packet is not dropped.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: Revert "ipsec: Use 64bits for XFRM output sequence number"

This reverts commit 4ea12ee32e48ea08de72aa36d45f8df0f5f468bf.

Revert ESN support, the initial implementation is incomplete because it
also needs to sync sequence numbers across restarts and node joins. For
now solve this problem by making key rotation painless and simple
through cilium CLI.

We can add ESN back once sequence number sync operation is in place.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

joestringer commented on pkg/datapath/loader/base.go:
```diff
@@ -408,6 +410,17 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+				if links, err := netlink.LinkList(); err == nil {
+					for _, link := range links {
+						linkName := link.Attrs().Name
+						if linkName != iface && strings.Contains(linkName, "eth*") {
```
`eth` is distro-specific, many distros use the new style of device naming scheme based on physical system layouts (en*) 

We hit an issue like that and resolved in https://github.com/cilium/cilium/pull/13234 , but in that case we were only cocerned with the device with the default route, not all devices.

jrfastab replied:
The other logic I considered is to use driver attached to the device, but that would also likely require cases for eks, ubuntu, etc.

jrfastab replied:
Or we could learn it from the source routes I guess which is probably the most reliable.

joestringer replied:
Any particular reason we can't make this consistent with the interface list detection introduced in the earlier commits of this tree? Just that it's trying to pull the "latest" set of all interfaces at Reinitialize time in case the device list changes during Cilium operation?

joestringer replied:
FWIW since I understand this is "additive", ie we try to append a set of interfaces to a list that already exists and could otherwise be populated by the user, I don't think this is necessarily a blocker for fixing this... but at the same time, if we're taking the effort to simplify these steps for other distros, then it makes sense for us to also follow up on Ubuntu since that's one of the most common distros used in cloud.

Last time we made an assumption about `eth0`, we found out within a week, maybe even same day that it was a problem.

joestringer replied:
Discussed out-of-band, should be good enough to have an exclusion list for virtual device types (veth, ipvlan, tunnel?) and select any other devices.

jrfastab replied:
Updates here, https://github.com/cilium/cilium/pull/15357/commits/3deb07c7a556e9b7accb24fafe428daf0d4ee90e

jrfastab replied:
also moved it into the initial config handling so its not dangling off in setup code.

joestringer commented on pkg/option/config.go:
```diff
@@ -2560,7 +2560,7 @@ func (c *DaemonConfig) Populate() {
 	c.EnableHostFirewall = viper.GetBool(EnableHostFirewall)
 	c.EnableLocalRedirectPolicy = viper.GetBool(EnableLocalRedirectPolicy)
-	c.EncryptInterface = viper.GetString(EncryptInterface)
+	c.EncryptInterface = viper.GetStringSlice(EncryptInterface)
```
Does this impact upgrade for users already specifying `EncryptInterface`? Let's say you deployed using the instructions in the docs today and upgrade to this PR. What happens?

Related, this daemon option doesn't live in a box by itself, it may be populated from Helm and the docs here describe how to do so:
https://docs.cilium.io/en/v1.9/gettingstarted/encryption/#encryption-interface

These would also need updating, but also if we're changing the instructions then we need to consider how it affects users that upgrade and release-note it assuming we're OK with the user impact.

jrfastab replied:
I believe it does not have an upgrade issue. As far as I understand it a string should be parsed the same as a string slice considering interface names can not have spaces.

joestringer commented on pkg/datapath/linux/route/route_linux.go:
```diff
@@ -514,6 +514,18 @@ func lookupDefaultRoute(family int) (netlink.Route, error) {
+	}
+
+	for _, route := range routes {
+		netlink.RouteDel(&route)
```
Is this really error-free? Seems like we shouldn't be ignoring the error here, even if it's a rare case like some administrator messing with route tables while we run this code from Cilium.

joestringer replied:
^ Probably good enough just to log it here at Warning level and proceed, since overall we just want to make the best effort to delete all of the relevant routes/table and move on.

joestringer replied:
Maybe one more level of smartness, if we can do an `errors.Is(err, unix.ENOENT)`  then happy to ignore the error. If it's *not* ENOENT then I think we should be more concerned since the route wasn't deleted.

jrfastab replied:
Pushes a warning now if route was not deleted. I'm not sure when a ENENT would happen her we are walking the we just fetched.

joestringer commented on pkg/datapath/linux/route/route_linux.go:
```diff
@@ -514,6 +514,18 @@ func lookupDefaultRoute(family int) (netlink.Route, error) {
+	for _, route := range routes {
+		netlink.RouteDel(&route)
+	}
+	return nil
```
This function is named `DeleteRouteTable`, but it appears to delete the rules *in* that table. I'm not familiar enough with the kernel API to say whether the table itself is a resource that needs to be deleted, but for readability would it be more accurate to call it something like `DeleteRoutesInTable(...)`? (or do we need an actual table delete call?)

jrfastab replied:
Looks like the kernel just adds and deletes tables as needed.

https://elixir.bootlin.com/linux/v4.19.182/source/net/ipv4/fib_frontend.c#L1424

joestringer commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -385,27 +385,33 @@ func DeleteIPsecEndpoint(remote *net.IPNet) {
 
 func isXfrmPolicyCilium(policy netlink.XfrmPolicy) bool {
 	if policy.Mark == nil {
+		// Check if its our fwd rule, we don't have a mark
```
FWIW this commit seems like it could do with a unit test. I don't know how reasonable that is in the immediate term but would be good to note down at least.

jrfastab replied:
Do our unit tests run in network namespaces though? Otherwise we will be stomping on the host most likely.

joestringer replied:
We have some tests where we set up the netns as part of the setup/teardown for isolation, such as those in `pkg/datapath/linux/node_linux_test.go`.

joestringer commented on pkg/datapath/linux/node.go:
```diff
@@ -841,6 +841,13 @@ func (n *linuxNodeHandler) enableIPsec(newNode *nodeTypes.Node) {
+				/* Insert wildcard policy rules for traffic skipping back through host */
+				ipsecIPv4Wildcard := &net.IPNet{IP: net.ParseIP(wildcardIPv4), Mask: net.IPv4Mask(0, 0, 0, 0)}
+				if err = ipsec.IpSecReplacePolicyFwd(ipsecIPv4Wildcard, ipsecRemote, ipsecLocal, ipsecRemote); err != nil {
+					//if err = ipsec.IpSecReplacePolicyFwd(ipsecRemote, ipsecIPv4Wildcard, ipsecLocal, ipsecRemote); err != nil {
```
extraneous line from development/debugging?

jrfastab replied:
oops yep I originally wrote it backwards and forgot to remove thanls

joestringer commented on pkg/datapath/loader/compile.go:
```diff
@@ -57,6 +57,12 @@ const (
+
+	networkPrefix   = "bpf_network"
+	networkProg     = networkPrefix + "." + string(outputSource)
+	networkObj      = networkPrefix + ".o"
```
If these are purely for documentation, we can add them as comments to work around the linter. If you intended to use them, then by all means let's do that ;)

jrfastab replied:
can be removed

joestringer reviewed (commented):
Mostly minor nits, just a couple of points on upgrade and Ubuntu support.

aditighag commented on daemon/cmd/daemon_main.go:
```diff
@@ -1258,12 +1259,30 @@ func initEnv(cmd *cobra.Command) {
+
+			if links, err := netlink.LinkList(); err == nil {
+				for _, link := range links {
+					use, err := ethtool.IsNotVirtualDriver(link.Attrs().Name)
```
Looks like this only excludes `veth`* devices? How about `cilium_host` interface?

jrfastab replied:
cilium_host is a veth device,
```
# ethtool -i cilium_host
driver: veth
version: 1.0
firmware-version: 
expansion-rom-version: 
bus-info: 
supports-statistics: yes
supports-test: no
supports-eeprom-access: no
supports-register-dump: no
supports-priv-flags: no
```

aditighag commented on daemon/cmd/daemon_main.go:
```diff
@@ -1258,12 +1259,30 @@ func initEnv(cmd *cobra.Command) {
+						interfaces = append(interfaces, link.Attrs().Name)
+					}
+				}
+				option.Config.EncryptInterface = interfaces
```
We are overriding user-specified config option here, no? I wonder if this can lead to confusion. At the very least, it'll be good to document this?

jrfastab replied:
If the user set an interface we skip this entire block.
```
        if option.Config.EnableIPSec && option.Config.Tunnel == option.TunnelDisabled && len(option.Config.EncryptInterface) == 0 {
                // IPAMENI mode supports multiple network facing interfaces that
                // will all need Encrypt logic applied in order to decrypt any
                // received encrypted packets. This logic will attach to all
                // !veth devices.
                if option.Config.IPAM == ipamOption.IPAMENI {
                        interfaces := option.Config.EncryptInterface

                        if links, err := netlink.LinkList(); err == nil {
                                for _, link := range links {
                                        use, err := ethtool.IsNotVirtualDriver(link.Attrs().Name)
                                        if err == nil && use {
                                                interfaces = append(interfaces, link.Attrs().Name)
                                        }
                                }
                                option.Config.EncryptInterface = interfaces
                        }
                } else {
                        link, err := linuxdatapath.NodeDeviceNameWithDefaultRoute()
                        if err != nil {
                                log.WithError(err).Fatal("Ipsec default interface lookup failed, consider \"encrypt-interface\" to manually configure interface.")
                        }
                        option.Config.EncryptInterface = append(option.Config.EncryptInterface, link)
                }
        }
```

aditighag replied:
In the ENI mode, users could specify incorrect interface which can lead to packets not getting decrypted. Any reason why we don't want to enable auto-detect for interfaces in such environments regardless of the user-specified config?

jrfastab replied:
I think if the user adds interfaces we should do what they tell us. It gives us an escape route if this breaks for whatever reason.

aditighag reviewed (commented):
I mainly reviewed the "auto-detect network interfaces " commit.

aditighag reviewed (approved):
Commit "auto-detect network interfaces" changes look fine to me. I'll leave the decision to override `encryption.interface` config option for IPAM ENI environments or not to John.

joestringer commented on pkg/datapath/linux/ethtool/ethtool.go:
```diff
@@ -0,0 +1,83 @@
+
+	switch drvName {
+	case veth:
+		return false, nil
```
Should we also include a variety of other ones like vxlan, geneve, ipvlan here?

jrfastab replied:
At the moment its guarded by ENI and !tunnel so its likely complete for the current use case. I guess its a toss up if we want to expand in this PR or in a follow up to extract this auto-detection out to other modes.

joestringer commented on pkg/datapath/loader/base.go:
```diff
@@ -418,6 +392,27 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+						option.Config.IPv4PodSubnets = append(option.Config.IPv4PodSubnets, &c)
+					}
+				}
+				log.Info("Encryption IPv4PodSubnets in-use")
```
This should be printed once on startup at most right? Not once every time we reinitialize the datapath? (Or at Debug level?)

jrfastab replied:
Lets just remove it this is a bit pointless.

joestringer commented on pkg/datapath/loader/base.go:
```diff
@@ -418,6 +392,27 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
 
+	// Compile and load network programs, currently used for encryption.
+	if option.Config.EnableIPSec {
+		interfaces := option.Config.EncryptInterface
```
I realized that I commented about converging this with the initial detection during daemon startup, but what happens if new ENI devices are added at runtime? Is the new version here properly detecting & applying the network progs to devices dynamically now?

jrfastab replied:
Can you add ENI devices at runtime without restarting the CNI? If so then we need to move it back.

twpayne replied:
Yes, ENI devices get added dynamically. Main reason is that a single ENI only supports a few IP addresses (e.g. 8-ish) so new ENIs need to be created and attached as the number of pods on the node grows.

jrfastab replied:
OK so we need to move it back to its original location I guess :/

joestringer reviewed (commented):
New version looks more robust for other distros. I realize one of my comments from last review might mess up the case where EKS environments need to add new ENI devices though, I don't think they're static enough that simply detecting the right set of devices on startup is sufficient.

jrfastab committed:
cilium: clean up bpf_network includes

bpf_network is pulling in unneeded headers resulting in compile errors
on older kernels. Use the linux/*.h headers for the handful of types
that the program actually uses.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

pchaigno commented on pkg/datapath/linux/datapath.go:
```diff
@@ -27,8 +27,8 @@ import (
 	HostDevice string
-	// EncryptInterface is the name of the device to be used for direct ruoting encryption
-	EncryptInterface string
+	// EncryptInterfaces is the list of device to be used for direct ruoting encryption
```
```suggestion
	// EncryptInterfaces is the list of devices used for direct routing encryption
```

pchaigno commented on pkg/datapath/linux/datapath.go:
```diff
@@ -52,9 +52,9 @@ func NewDatapath(cfg DatapathConfiguration, ruleManager datapath.IptablesManager
-			log.WithField(logfields.Interface, cfg.EncryptInterface).Warn("Rpfilter could not be disabled, node to node encryption may fail")
+	for _, iface := range cfg.EncryptInterfaces {
+		if err := connector.DisableRpFilter(iface); err != nil {
+			log.WithField(logfields.Interface, iface).Warn("Rpfilter could not be disabled, node to node encryption may fail")
```
Should we also log the error (`WithError(err)`) here?

jrfastab replied:
yes

pchaigno commented on pkg/datapath/linux/node.go:
```diff
@@ -459,8 +459,17 @@ func upsertIPsecLog(err error, spec string, loc, rem *net.IPNet, spi uint8) {
+func getDefaultEncryptionInterface() string {
+	iface := ""
+	if len(option.Config.EncryptInterface) > 0 {
+		iface = option.Config.EncryptInterface[0]
```
We need to document the special meaning of the first interface in the list.

jrfastab replied:
This works for EKS because it doesn't seem to care what interface we use. I was going to fix this up in another series to do this properly after I decide what that means. I think it means doing a route lookup to find the interface instead of encoding it here. But for this series trying to minimize changes to get to minimal good state on eks.

jrfastab replied:
I added a comment to the code, but it feels a bit lame. Anyone want to object loudly to pushing the further cleanup/fix to a follow up PR. I'm a bit hesitant to start doing major refactors here until I have CI in place to test them on the various envs (EKS, AKS, GKE, etc.) and can't get those down until this PR is merged.

pchaigno commented on pkg/datapath/loader/loader.go:
```diff
@@ -369,6 +370,18 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
+	}
+	for _, iface := range option.Config.EncryptInterface {
+		if err := l.replaceDatapath(ctx, iface, networkObj, symbolFromNetwork, dirIngress); err != nil {
+			log.Fatalf("Load encryption network (%s) failed", iface)
```
```suggestion
			log.WithField(logfields.Interface, iface).Fatal("Load encryption network failed")
```

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -1259,11 +1260,29 @@ func initEnv(cmd *cobra.Command) {
+
+			if links, err := netlink.LinkList(); err == nil {
+				for _, link := range links {
+					use, err := ethtool.IsNotVirtualDriver(link.Attrs().Name)
```
The commit message needs to be updated. It still says:
```
For an initial effort just pattern match the EKS case and
attach to any network interfaces with eth* in the name.
```
but it looks like it was updated to select all interfaces except for veth.

jrfastab replied:
correct. Patch was updated but description was not.

pchaigno commented on pkg/datapath/linux/node.go:
```diff
@@ -1078,6 +1078,10 @@ func (n *linuxNodeHandler) removeEncryptRules() error {
 	}
 
+	if err := route.DeleteRouteTable(linux_defaults.RouteTableIPSec, netlink.FAMILY_V4); err != nil {
+		log.WithError(err).Warn("DeleteRouteTable failed")
```
```suggestion
		log.WithError(err).Warn("Deletion of IPSec routes failed")
```

pchaigno commented on pkg/datapath/linux/route/route_linux.go:
```diff
@@ -514,6 +514,24 @@ func lookupDefaultRoute(family int) (netlink.Route, error) {
+	for _, route := range routes {
+		err := netlink.RouteDel(&route)
+		if err != nil {
+			routeErr = fmt.Errorf("DeleteRouteTable ReouteDel error: %s", err)
```
```suggestion
			routeErr = fmt.Errorf("Failed to delete route: %s", err)
```

pchaigno commented on pkg/datapath/loader/compile.go:
```diff
@@ -397,3 +406,37 @@ func compileTemplate(ctx context.Context, out string, isHost bool) error {
+		Output:  option.Config.StateDir,
+		State:   option.Config.StateDir,
+	}
+	scopedLog := log.WithField(logfields.Debug, true)
```
What is this line doing?

jrfastab replied:
Its creating the logger to pass into the cmd below so any errors are logged into the logger instead of stderr. Does that help?

pchaigno replied:
I understand that part. I'm wondering what's up with `WithField(logfields.Debug, true)`. Does that mean all subsequent messages are debug logs?

jrfastab replied:
@joestringer do you know? I just copied the other users here.

joestringer replied:
Honestly not sure, at best it seems misleading given that there are some non-debug logs lower in the function; at worst it's hiding logs that should really be hitting the log.

pchaigno reviewed (commented):
Only minor comments below.

I'm having a hard time reviewing these two commits though:
```
cilium: Use correct mask on xfrm delete
cilium: encryption, insert "skip" rules to avoid ipsec policy 
```
They could probably use an example. I'm not sure if someone else feels they have enough context to review those.

jibi commented on pkg/datapath/linux/ethtool/ethtool.go:
```diff
@@ -0,0 +1,83 @@
+package ethtool
```
This is missing the copyright header

jrfastab replied:
added.

jrfastab replied:
```
@@ -1,3 +1,17 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```

jibi commented on pkg/datapath/linux/config/config.go:
```diff
@@ -419,8 +419,11 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
-	if iface := option.Config.EncryptInterface; iface != "" {
-		link, err := netlink.LinkByName(iface)
+	if iface := option.Config.EncryptInterface; len(iface) != 0 {
+		// When FIB lookup is not supported we need to pick an interface
```
I'm not sure I understand this comment, why is that `len(iface) != 0` => `FIB lookup not supported`?

jrfastab replied:
I can improve comment I suppose, the reason is if FIB lookup is supported the datapath will do a lookup to find the egress device, but if its not support on the kernel we tell the datapath what interface to use.

jibi commented on pkg/datapath/linux/ethtool/ethtool.go:
```diff
@@ -0,0 +1,83 @@
+	return nil
+}
+
+func GetDeviceName(iface string) (string, error) {
```
Nit: I think this should be called `GetDriverName`? Also I see we have already `github.com/safchain/ethtool` vendored in, which exposes a [similar method](https://github.com/safchain/ethtool/blob/master/ethtool.go#L238), might be worth using that?

jrfastab replied:
I don't think we are using safchain/ethtool anywhere in Cilium, at least not in ./pkg, and would prefer to avoid pulling in another dependency if possible. Seems best to leave it here IMO.

joestringer replied:
I'm inclined to agree with John here. I don't know why it's in go.sum, we don't use it anywhere:

```
$ git grep safchain
go.sum:github.com/safchain/ethtool v0.0.0-20190326074333-42ed695e3de8/go.mod h1:Z0q5wiBQGYcxhMZ6gUqHn6pYNLypFAvaL3UvgZLR0U4=
```

twpayne commented on pkg/datapath/linux/ethtool/ethtool.go:
```diff
@@ -0,0 +1,83 @@
+	return string(bytes.Trim(info.driver[:], "\x00")), nil
+}
+
+func IsNotVirtualDriver(iface string) (bool, error) {
```
nit: I'd flip this function use positive logic (`IsVirtualDriver`) as it makes it easier to reason about.

twpayne commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -386,27 +386,33 @@ func DeleteIPsecEndpoint(remote *net.IPNet) {
 	}
-	if policy.Mark.Value == linux_defaults.RouteMarkDecrypt ||
-		policy.Mark.Value == linux_defaults.RouteMarkEncrypt {
+	if (policy.Mark.Value & linux_defaults.RouteMarkEncrypt) != 0 {
```
nit: These two if statements could be combined (it took me a moment to spot the difference between them):

```go
if policy.Mark.Value & (linux_defaults.RouteMarkDecrypt | linux_defaults.RouteMarkEncrypt) != 0 {
    return true
}
```

joestringer replied:
I'm not sure this is a great idea. The mark bits are not unique here, we're treating the `0x0F00` bits as a 4-bit value (to extract up to 15 unique values out of the bits rather than 4). We don't want to see whether bits `0x0F00` are set (which is what `0xD | 0xE` results in), we want to know the rule uses one of two specific values within that set of four bits.

jrfastab replied:
I think I agree with @joestringer here. I'm leaning towards being overly verbose.

twpayne commented on pkg/datapath/linux/route/route_linux.go:
```diff
@@ -514,6 +514,24 @@ func lookupDefaultRoute(family int) (netlink.Route, error) {
+	for _, route := range routes {
+		err := netlink.RouteDel(&route)
+		if err != nil {
+			routeErr = fmt.Errorf("Failed to delete route: %s", err)
```
Only the last error encountered here will be returned, others will be lost. Maybe log and/or combine multiple errors?

jrfastab replied:
ok combining makes the most sense to me.

jrfastab replied:
It might print a bit clumsy if we get a bunch of errors here, but hopefully that doesn't happen.

~  routeErr = fmt.Errorf("%w: Failed to delete route: %s", routeErr, err)

twpayne commented on pkg/datapath/loader/base.go:
```diff
@@ -177,6 +176,49 @@ func addENIRules(sysSettings []sysctl.Setting, nodeAddressing datapath.NodeAddre
 	return retSettings, nil
 }
 
+// reinitializeIPSec is used to recompile and load encryption network porgrams.
```
nit: s/porgrams/programs/.

jrfastab replied:
fixed

twpayne commented on pkg/datapath/loader/base.go:
```diff
@@ -177,6 +176,49 @@ func addENIRules(sysSettings []sysctl.Setting, nodeAddressing datapath.NodeAddre
+		if len(interfaces) == 0 {
+			if links, err := netlink.LinkList(); err == nil {
+				for _, link := range links {
+					use, err := ethtool.IsNotVirtualDriver(link.Attrs().Name)
```
nit: use `!ethtool.IsVirtualDriver(...)` (see previous nit).

jrfastab replied:
Yes fixed up with 
```
isVirtualDevice, err := ethtool.IsVirtualDriver(...)
if !isVirtualDevice ...
```

twpayne commented on pkg/option/config.go:
```diff
@@ -1383,7 +1383,7 @@ type DaemonConfig struct {
 	HostV4Addr          net.IP     // Host v4 address of the snooping device
 	HostV6Addr          net.IP     // Host v6 address of the snooping device
-	EncryptInterface    string     // Set with name of network facing interface to encrypt
+	EncryptInterface    []string   // Set of network facing interface to encrypt over
```
Is it worth renaming this variable to `EncryptInterfaces` to better reflect its new type?

joestringer replied:
I think this re-raises the upgrade question which we earlier managed to convince ourselves was not too much of a concern. Given this is targeted as a fix for backport, I think we should try to limit such changes, and instead target such a change for a future minor (X.Y.0) release.

jrfastab replied:
ah the problem is if we do that we break backwards compatibility. Right now old configs will continue to work because a single string will be read as a len(EncryptInterface) == 1 slice. And new users can put a 'space' separated list in there and it will be consumed as a slice. If we rename it that breaks. I guess we could rename just the value in the struct and tell the CLI to use a different name and map them, but that seems even more confusing.

pchaigno replied:
Instead of renaming we could just introduce a new `EncryptInterfaces` flag and slowly deprecate `EncryptInterface`.

twpayne reviewed (approved):
LGTM, just a few minor nits.
